lake/commit/snapshot.go

```diff
@@ -1,48 +1,112 @@
 package commit
 
 import (
+	"errors"
+
+	"github.com/brimdata/zed/lake/commit/actions"
 	"github.com/brimdata/zed/lake/journal"
 	"github.com/brimdata/zed/lake/segment"
 	"github.com/brimdata/zed/pkg/nano"
-	"github.com/brimdata/zed/zbuf"
 	"github.com/segmentio/ksuid"
 )
 
+var (
+	ErrExists   = errors.New("segment exists")
+	ErrNotFound = errors.New("segment not found")
+)
+
+type View interface {
+	Lookup(id ksuid.KSUID) (*segment.Reference, error)
+	Select(span nano.Span) Segments
+}
+
+type Writeable interface {
+	View
+	AddSegment(seg *segment.Reference) error
+	DeleteSegment(id ksuid.KSUID) error
+}
+
 // A snapshot summarizes the pool state at a given point in the journal.
-// XXX TBD: sort log by span and stack snapshot as a base layer
-// followed by change sets for each base-layer generation.  This is what
-// we will store in the journal sub-pool when this gets implemented.
-// See issue #XXX.
-// Also, snapshots should have index updates so each segment in the
-// snapshot could include which indexes are attached to it.
 type Snapshot struct {
 	at       journal.ID
-	order    zbuf.Order
-	segments []segment.Reference
+	segments map[ksuid.KSUID]*segment.Reference
 }
 
-func newSnapshot(at journal.ID, order zbuf.Order) *Snapshot {
-	return &Snapshot{at: at, order: order}
+func NewSnapshot() *Snapshot {
+	return &Snapshot{
+		segments: make(map[ksuid.KSUID]*segment.Reference),
+	}
+}
+
+func newSnapshotAt(at journal.ID) *Snapshot {
+	s := NewSnapshot()
+	s.at = at
+	return s
 }
 
-func (s *Snapshot) Segments() []segment.Reference {
-	return s.segments
+func (s *Snapshot) AddSegment(seg *segment.Reference) error {
+	id := seg.ID
+	if _, ok := s.segments[id]; ok {
+		return ErrExists
+	}
+	s.segments[id] = seg
+	return nil
 }
 
-func (s *Snapshot) AddSegment(seg segment.Reference) {
-	s.segments = append(s.segments, seg)
+func (s *Snapshot) DeleteSegment(id ksuid.KSUID) error {
+	if _, ok := s.segments[id]; !ok {
+		return ErrNotFound
+	}
+	delete(s.segments, id)
+	return nil
 }
 
-func (s *Snapshot) DeleteSegment(id ksuid.KSUID) {
-	panic("TBD")
+func (s *Snapshot) Exists(id ksuid.KSUID) bool {
+	_, ok := s.segments[id]
+	return ok
 }
 
-func (s *Snapshot) Select(span nano.Span) []*segment.Reference {
-	var segments []*segment.Reference
-	for k, seg := range s.segments {
+func (s *Snapshot) Lookup(id ksuid.KSUID) (*segment.Reference, error) {
+	seg, ok := s.segments[id]
+	if !ok {
+		return nil, ErrNotFound
+	}
+	return seg, nil
+}
+
+func (s *Snapshot) Select(span nano.Span) Segments {
+	var segments Segments
+	for _, seg := range s.segments {
 		if span.Overlaps(seg.Span()) {
-			segments = append(segments, &s.segments[k])
+			segments = append(segments, seg)
 		}
 	}
 	return segments
 }
+
+type Segments []*segment.Reference
+
+func (s *Segments) Append(segments Segments) {
+	*s = append(*s, segments...)
+}
+
```

Comment:
Is this type going to get more methods? If not, it should go away.

Reply:
I think it will change when we put the spanshots into a sub-pool.  I created the type to abstract things so there will be fewer changes down the road when we make this change.